### PR TITLE
Fix connectRange & RangeInput

### LINKED ISSUE #428: connectRange ignore boundaries when refine
When boundaries are set the user is still able to select a range outside of it.

**Step to reproduce with `RangeInput`:**

* set the max boundary to `300`
* set the max input to `450`
* look at the query parameters you will have `"price<=450"` instead of `price<=300`

**Possible solution:**

We have to handle the logic in `connectRange` to avoid input values to go above the limit (for user that use the connector API). But we also have to handle the limit in the `RangeInput` component to avoid end user to go above the limit in the input.

[Storybook example](https://community.algolia.com/react-instantsearch/storybook/?knob-max=300&knob-translations=%7B%22submit%22%3A%22%20go%22%2C%22separator%22%3A%22to%22%7D&selectedKind=RangeInput&selectedStory=playground&full=0&down=1&left=1&panelRight=1&downPanel=storybooks%2Fstorybook-addon-knobs)

### PR DESCRIPTION
**Pull request by samouss**
**Summary**

*Sorry for the pretty long PR, I merged another fix into it and I shouldn't have.*

Fix #428 #429 #430

This PR aims to fix some problems related to `connectRange` & `RangeInput`. 

For `connectRange` improvements are:

- you can now reset the refinement
- you cannot refine with values outside of the range
- you can now specify the number of digits for decimal values with `precision` option (same as IS)

For `RangeInput` improvements are:

- the range is displayed with placeholders (same as IS)
- you can now reset the refinement (same as IS)
- you can refine with decimal value with the `precision` option (same as IS)
- you cannot refine out of bounds, thanks to the HTML5 validation (same as IS)

**Result**

You can play with the `RangeInput` on Storybook. You can also play with the IS version of it for compare the behaviour of both, you can access to it [here](https://deploy-preview-2440--algolia-instantsearch.netlify.com/v2/dev-novel/?selectedStory=RangeInput.default).

### CODE REVIEW

File: stories/RangeInput.stories.js [hunk 1]
```diff
@@ -18,6 +18,26 @@ stories
...
+      <div style={{ display: 'none' }}>
+        <SearchBox defaultRefinement="ds" />
+      </div>
+    </WrapWithHits>
+  ))
+  .add('with precision of 2', () => (
+    <WrapWithHits linkedStoryGroup="RangeInput">
+      <RangeInput attributeName="price" precision={2} />
+    </WrapWithHits>
+  ))
```
Comment (mthuret) [inline stories/RangeInput.stories.js]: You can add `precision` as a field of the `playground` story

File: packages/react-instantsearch/src/connectors/connectRange.js [hunk 2]
```diff
@@ -76,29 +77,86 @@ function getCurrentRefinement(props, searchState, context) {
...
-function refine(props, searchState, nextRefinement, context) {
-  if (
-    !_isFinite(parseFloat(nextRefinement.min)) ||
-    !_isFinite(parseFloat(nextRefinement.max))
-  ) {
-    throw new Error(
-      "You can't provide non finite values to the range connector"
-    );
+function refine(props, searchState, nextRefinement, currentRange, context) {
+  const { min: nextMin, max: nextMax } = nextRefinement;
```
Comment (mthuret) [inline packages/react-instantsearch/src/connectors/connectRange.js]: maybe we could create one or several intermediary functions?

File: package.json [hunk 3]
```diff
@@ -131,7 +132,8 @@
     ],
     "setupFiles": [
       "./shim.js"
-    ]
+    ],
+    "snapshotSerializers": ["enzyme-to-json/serializer"]
```
Comment (mthuret) [inline package.json]: what does that bring?

**Comment by mthuret** [comment]
I forget what `reset` the refinement means, could explain? :)

File: package.json (hunk 3 above)
Comment (Haroenv) [reply package.json]: It should make sure that the snapshots made with enzyme methods like shallow are also looking JSX, like the raw jsx would be I think

see [yarn.pm/jest-serializer-enzyme](https://yarn.pm/jest-serializer-enzyme)

Comment (samouss) [reply package.json]: @Haroenv says right (the original package is [this one](https://github.com/adriantoine/enzyme-to-json))

It allows you to do:

```js
import { shallow } from 'enzyme';

const wrapper = shallow(MyComponent);

expect(wrapper).toMatchSnapshot();
```

We don't rely on `react-test-renderer` inside the test anymore. You can handle all your tests on components with enzyme (Snapshot + assertions).

**Comment by samouss** [comment]
The `reset` refinement means that you can undo your refinement. In the current implementation if you refine with `10` you can't reset the refinement via the input. You need to pass via the `Clear` refinements button. Now you can just reset the input.

**Before:**

before

**After:**

after

**Comment by mthuret** [comment]
Hey @samouss could you check the argos differences? 

I spotted two I have questions about: 

- we now display 0 instead 0.39, is it expected? 
- there's also a difference with the slider, is that due to the precision modification?

**Comment by samouss** [comment]
@mthuret Yep the difference on Argos are due to the precision modification. The default precision is `0` for share the same behaviour as `InstantSearch`. That's why the input is round to `0` instead to `0.39`. Same for the slider the changes is due to range that is bit longer than before. If we want to share the same behaviour as before, we just need to set the default precision to `2`.

**Comment by mthuret** [comment]
@samouss I would go for keeping the same behaviour as before for now as I don't want to bump the version yet. 

However we can open an issue and handle this later for a v5. What do you think?

**Comment by mthuret** [comment]
Some precision: when I say later, it'll be indeed part of the InstantSearch.css compliance version :)

**Comment by samouss** [comment]
@mthuret Should be good now! The differences on Argos are related to the changes on the range. We don't display the `min` & `max` as value now but as `placeholder`. The behaviour for the query is the same.

**Pull request merged**